src/bookmarks/ephy-bookmarks-manager.c

```diff
@@ -611,3 +624,294 @@ ephy_bookmarks_manager_save_to_file_warn_on_error_cb (GObject      *object,
+                                                       EPHY_PREFS_SYNC_WITH_FIREFOX);
+
+  return sync_with_firefox ? "bookmarks" : "ephy-bookmarks";
+}
+
+static GType
+synchronizable_manager_get_synchronizable_type (EphySynchronizableManager *manager)
+{
+  return EPHY_TYPE_BOOKMARK;
+}
+
+static gboolean
+synchronizable_manager_is_initial_sync (EphySynchronizableManager *manager)
+{
+  return g_settings_get_boolean (EPHY_SETTINGS_SYNC,
+                                 EPHY_PREFS_SYNC_BOOKMARKS_INITIAL);
+}
+
+static void
+synchronizable_manager_set_is_initial_sync (EphySynchronizableManager *manager,
+                                            gboolean                   is_initial)
+{
+  g_settings_set_boolean (EPHY_SETTINGS_SYNC,
+                          EPHY_PREFS_SYNC_BOOKMARKS_INITIAL,
+                          is_initial);
+}
+
+static double
+synchronizable_manager_get_sync_time (EphySynchronizableManager *manager)
+{
+  return g_settings_get_double (EPHY_SETTINGS_SYNC,
+                                EPHY_PREFS_SYNC_BOOKMARKS_TIME);
+}
+
+static void
+synchronizable_manager_set_sync_time (EphySynchronizableManager *manager,
+                                      double                     sync_time)
+{
+  g_settings_set_double (EPHY_SETTINGS_SYNC,
+                         EPHY_PREFS_SYNC_BOOKMARKS_TIME,
+                         sync_time);
+}
+
+static void
+synchronizable_manager_add (EphySynchronizableManager *manager,
+                            EphySynchronizable        *synchronizable)
+{
+  EphyBookmarksManager *self = EPHY_BOOKMARKS_MANAGER (manager);
+  EphyBookmark *bookmark = EPHY_BOOKMARK (synchronizable);
+
+  ephy_bookmarks_manager_add_bookmark_internal (self, bookmark, TRUE);
+  ephy_bookmarks_manager_create_tags_from_bookmark (self, bookmark);
+}
+
+static void
+synchronizable_manager_remove (EphySynchronizableManager *manager,
+                               EphySynchronizable        *synchronizable)
+{
+  EphyBookmarksManager *self = EPHY_BOOKMARKS_MANAGER (manager);
+  EphyBookmark *bookmark = EPHY_BOOKMARK (synchronizable);
+
+  ephy_bookmarks_manager_remove_bookmark_internal (self, bookmark);
+}
+
+static GSList *
+ephy_bookmarks_manager_handle_initial_merge (EphyBookmarksManager *self,
+                                             GSList               *remote_bookmarks)
+{
+  GSList *to_upload = NULL;
+  EphyBookmark *bookmark;
+  GSequence *bookmarks;
+  GSequenceIter *iter;
+  GHashTable *dont_upload;
+  double timestamp;
+
+  g_assert (EPHY_IS_BOOKMARKS_MANAGER (self));
+
+  dont_upload = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+
+  for (GSList *l = remote_bookmarks; l && l->data; l = l->next) {
+    const char *id;
+    const char *url;
+    char *type;
+    char *parent_id;
+
+    g_object_get (l->data, "type", &type, "parentid", &parent_id, NULL);
+    /* Ignore mobile/unfiled bookmarks and everything that is not of type bookmark. */
```

Comment:
Can we display them under a Mobile tag, similar to how Firefox does? We definitely want to be able to sync bookmarks created on the phone to the desktop. I think implementing the same behavior as Firefox seems fine.

But if desktop Firefox does not display those bookmarks, then I guess we can explain it away as expected behavior.